Q1: The majority of CpG dinucleotides are methylated.

3a:
Percentage of shared sites: 95.85468138657977%
Percentage of nanopore unique sites: 2.9552115872243195%
Percentage of bisulfite unique sites: 1.1901070261959185%

Q2: Nanopore has a lower average coverage depth but bisulfite sequencing has a much higher proportion of sites with very low coverage depth. I think the nanopore methylation calling is better because you have fewer low confidence methylation sites.

Q3: Oxford nanopore appears to be a better method for detecting changes in methylation as it contains fewer sites with a difference close to 0 

Q4: Globally there is a bias towards an increase in methylation.

Q5: There is an increase in methylation in the tumor sample within the gene. This indicates silencing of the gene which could lead to loss of methylation overtime as it normally serves to preserve methylation after DNA replication by recognizing hemimethylated sites.

Q6: It means that its expression has been modified by epigenetic marks established in the germline of the parents.

Q7: The reads are being grouped based on the allele that they come from. You need sufficient read length to identify reads coming from one allele or the other.

Q8: No, not every set of reads can be phased. There needs to be enough differences between the two alleles to be able to identify where the reads came from. You also need to have long enough reads to avoid ambiguity.